src/pb_stub.cc

```diff
@@ -870,6 +879,59 @@ Stub::ProcessRequests(RequestBatch* request_batch_shm_ptr)
   }
 }
 
+py::object
+Stub::GetAsyncEventLoop()
+{
+  if (py::isinstance<py::none>(async_event_loop_)) {
+    // Create the event loop if not already.
+    py::module asyncio = py::module_::import("asyncio");
+    async_event_loop_ = asyncio.attr("new_event_loop")();
+    asyncio.attr("set_event_loop")(async_event_loop_);
+    py::object py_thread =
+        py::module_::import("threading")
+            .attr("Thread")(
+                "target"_a = async_event_loop_.attr("run_forever"),
+                "daemon"_a = true);
+    py_thread.attr("start")();
+  }
+  return async_event_loop_;
+}
+
+py::object
+Stub::RunCoroutine(py::object coroutine)
+{
+  py::object loop = GetAsyncEventLoop();
+  py::object py_future = py::module_::import("asyncio").attr(
+      "run_coroutine_threadsafe")(coroutine, loop);
+
+  std::shared_ptr<std::future<void>> shared_future(new std::future<void>());
+  std::future<void> c_future = std::async(
+      std::launch::async, [this, shared_future, py_future]() mutable {
+        {
+          py::gil_scoped_acquire gil_acquire;
+          try {
+            py_future.attr("result")();
+          }
+          catch (const PythonBackendException& pb_exception) {
+            LOG_ERROR << pb_exception.what();
+          }
+          catch (const py::error_already_set& error) {
+            LOG_ERROR << error.what();
+          }
+          py_future = py::none();
+        }
+        std::vector<std::shared_ptr<std::future<void>>> empty;
+        {
+          std::lock_guard<std::mutex> lock(async_event_futures_mu_);
+          done_async_event_futures_.swap(empty);
+          done_async_event_futures_.emplace_back(std::move(shared_future));
+        }
+      });
+  *shared_future = std::move(c_future);
```

Comment:
could you elaborate why we need a separate thread for every future that just waits for the future is C++?

Reply:
Since we want to enable overlapping in calls to async execute function and the function does not return anything meaningful (always return `None`), the stub process will report the execution is completed while leaving the async execute coroutine to execute in the background and allowing it to generate new responses until it returns. Two things can happen while the coroutine is in the background
1. It returns normally - not a concern.
2. It raises an exception. 

The purpose of the c future is to keep track of the py future until its coroutine is done. Without the tracking, any exception raised by the py future will be lost. The c future is blocked on the `py_future.result()` call, until the py future is completed. If the py future raises an exception, the c future will capture it and prints it to the log.

Since `py_future.result()` is blocking, separate threads are necessary. Why not just use the `py_future.done()` check, which is non-blocking? The done check only returns `True` if the coroutine completes successfully (or cancelled) and does not raise an exception. The goal is to capture the exception, so the only alternative is `py_future.exception()` which is equally blocking as the `py_future.result()`, so `py_future.result()` is picked for its easy to use, as it reports  anything abnormal as an exception.

Reply:
Can the user utilize the ` loop.set_exception_handler(handle_exception)` to handle the exceptions? We could have a default one that logs the exceptions?

Reply:
I don't think the `def handle_exception(loop, context)` set by `loop.set_exception_handler(handle_exception)` will get called. For example,
```python
#!/usr/bin/env python3

import asyncio
import threading

def get_async_event_loop():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    t = threading.Thread(target=loop.run_forever, daemon=True)
    t.start()
    return loop

def custom_exception_handler(loop, context):
    print("----- Exception caught by custom_exception_handler -----")
    print(context)

def run_coroutine(loop, coroutine_):
    py_future = asyncio.run_coroutine_threadsafe(coroutine_, loop)

    try:
        py_future.result()
    except Exception as e:
        print("----- Exception caught from py_future -----")
        print(e)

async def raise_exception():
    print("async raise_exception() start")
    await asyncio.sleep(1)
    print("async raise_exception() raise exception")
    raise ValueError("dummy exception")
    print("async raise_exception() end")

if __name__ == "__main__":
    loop = get_async_event_loop()
    loop.set_exception_handler(custom_exception_handler)
    run_coroutine(loop, raise_exception())
```
```
$ python3 aio_exception_handler.py
async raise_exception() start
async raise_exception() raise exception
----- Exception caught from py_future -----
dummy exception
```
Note: `loop.run_forever()` is a blocking until `loop.stop()` is called.

Reply:
I see, that's strange. @kthui can we perhaps have a background thread that monitors a vector of errors and logs them if there are any?

Reply:
> have a background thread that monitors a vector of errors and logs them if there are any

The issue is how to get a vector of errors? There are two ways to get the error:
1. [`py_future.result()`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result) which raises an exception for any error; or
2. [`py_future.exception()`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.exception) which return the error, except raise exception if cancelled.

Unfortunately, both APIs are blocking, so there is a need for a thread to wait for each `py_future` until it is completed.
Fortunately, the threads are C++ threads, so any overhead should be minimal compare to those in Python.

There are only 3 APIs that are non-blocking, but non (or any combinations) of them can tell if there is an exception:
* [cancelled](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.cancelled)
* [running](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.running)
* [done](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.done)

The definition of `done` there is the coroutine completed running and either it is cancelled or completed successfully, which means a coroutine that raised an exception is not done, leaving us with the blocking options above.

Reply:
Thanks for super detailed explanation. Could we perhaps use `add_done_callback`?

Reply:
Sure. I still have no luck getting the exception logging to match exactly what we have from the others, but I could get it to print sufficient details, i.e.
```
...
E0408 21:43:36.182801 6016 pb_stub.cc:120] Traceback (most recent call last):
  File "/opt/tritonserver/qa/L0_backend_python/async_execute/models/async_execute_decouple/1/model.py", line 44, in execute
    self.raise_value_error(requests)
  File "/opt/tritonserver/qa/L0_backend_python/async_execute/models/async_execute_decouple/1/model.py", line 77, in raise_value_error
    raise ValueError("wait_secs cannot be negative")
ValueError: wait_secs cannot be negative

...
```